### ISSUE TITLE
Consider `these-core` ?

### ISSUE THREAD

**Issue by fosskers** [isomorphism/these#108]
`these` is a wonderful library, but has quite a large dependency footprint. Would you consider publishing a `these-core` with only the main data type and `base`-based typeclass instances? `these` could then depend on `these-core`, and expose the rest of its machinery.

I'm aware that `data-or` exists, but it is missing certain instances that a modern Haskeller would expect. I'm trying to get it patched as well, but the maintainer (and the repo!) might be "lost to time". I will update this thread if the `data-or` situation changes.

If manpower is an issue, I can put in the work myself to create `these-core`, and then link `these` to it. Thoughts?

**Comment by fosskers** [comment]
Update: the listed email address for the maintainer of `data-or` seems to no longer exist.

**Comment by phadej** [comment]
The dependency closure of `these` (with GHC-8.4.4) looks like:

![deps](https://user-images.githubusercontent.com/51087/57978756-e703c380-7a1b-11e9-8183-ce4e01c91bd8.png)

There are
- things up to `semigroupoids`
- a little of `lens` on top
- and somewhat separate `aeson` closure
- `assoc`
- and `QuickCheck`

I haven't worked on an application that wouldn't (directly or transitively) depend on `semigroupoids` nor `aeson`. Or don't use `QuickCheck` in its tests (ok, you might use `hedgehog` for your property based tests). All of these (pun intended) dependencies are well maintained: at least as well maintained as `these` itself. Bugs are fixed, bounds are relaxed etc in reasonable time.

`cabal v2-build`, `stack` and Nix are all good at caching dependencies. And these dependencies are quite stable, so caches aren't invalidated often. Also CIs can use caches.

Bottom line: I don't see a net benefit of splitting up `these` (or actually almost any package) for a reason of less dependency footprint.

---

However, I do consider splitting into separate packages based on **functionality**:

- `these`: `Data.These` and `Data.These.Combinators`
- `align` (that name is taken, ideas welcome): `Data.Align` (and maybe add `Data.Zip`, see https://github.com/ekmett/keys/issues/20)
- `these-lens`: `Data.These.Lens`
- and `monad-chronicle`: `Control.Monad.Chronicle` stuff

That might help with dependencies, though e.g. `these` would still depend on `aeson`, `assoc` and `QuickCheck`; `align` would depend on `vector-instances` (maybe, not clear with `keys` stuff) etc.

**Comment by fosskers** [comment]
Hi @phadej thanks for the reply. What command did you use to generate that dep graph output? `stack dot` gave me output [as seen here](https://github.com/mstksg/nonempty-containers/pull/2) (that PR also explains why I went for `data-or` in the first place - to exclude `lens` from the dep graph of a library).

That's the crux of it, I think: libraries probably have some duty to have as small a dependency graph as possible, since we don't know what our users expect w.r.t. transitive dependencies. Hence why `microlens` exists in the first place, and people often hand-write their lenses to avoid any extra deps (within libs).

The functionality-based split you describe sounds reasonable: it would make the transitive deps (and the functionality they bring) "opt-in". For `align`:

- `data-align`?
- `aligned`? 

For `these-lens`, since `lens` already depends on the world already, why not commit `Data.These.Lens` to `lens` directly?

`monad-chronicle` sounds good as its own library, and fits the pattern of the other MTL-style sublibs (like say `resourcet`).

Okay, so what would next steps be? Would you like my help with anything?

**Comment by emilypi** [comment]
I'm here because @fosskers and I work together and we're trying to work out a minimal dependency tree while still using `these`. The net benefit of splitting up the hierarchy in terms of, say, a trio of  `these`, `these-lens`, and `these-aeson`, means a great deal considering how many dependencies `lens` brings in. While we are subscribed to the Kmett universe of dependencies, if someone else were to just need _these_ and didn't have `lens` or `aeson` in scope, it would be dependency overkill. 

In our case, we do not have a dire need for you splitting up `these`, but it would make sense from a dependency hygiene perspective.

**Comment by fosskers** [comment]
I would personally still appreciate the split for usage in a large FOSS project that I have, which avoids the lens dep on purpose.

**Comment by phadej** [comment]
With #112 the default hierarchy looks like:

![deps](https://user-images.githubusercontent.com/51087/58754018-25e05180-84d1-11e9-8959-3832547cca6b.png)

However, I also added flags, so with `-assoc -QuickCheck -semigroupoids -aeson` you'll get

![deps-light](https://user-images.githubusercontent.com/51087/58754024-37c1f480-84d1-11e9-9b6e-a28cce626915.png)

I.e. `these` depends only on `base-compat` and `hashable` (`semialign` is quite light package too).

---

If this is not good enough, there seems to be https://hackage.haskell.org/package/these-skinny

**Comment by emilypi** [comment]
You're amazing thank you @phadej

**Comment by fosskers** [comment]
Thank you!

**Comment by fosskers** [comment]
These flags are on by default, was that intended?

**Comment by fosskers** [comment]
The flags seem on by default, was that intended?

**Comment by phadej** [comment]
@fosskers yes, and it won't change. Turning them off is on your own risk (e.g. if some dependency wants some instances).

**Comment by fosskers** [comment]
Gotcha, thanks again!

**Comment by phadej** [comment]
I noticed that `vector-instances` now unneeded dependency, so `these-lens` & `semialign-indexed` *only* add `lens` dependency. For ones who care.

![deps](https://user-images.githubusercontent.com/51087/58759228-548d1500-852f-11e9-8a90-b8eb9c162359.png)

![deps-light](https://user-images.githubusercontent.com/51087/58759234-65d62180-852f-11e9-9c66-48eb7785b674.png)
